sdks/python/apache_beam/runners/interactive/display/pcoll_visualization_test.py

```diff
@@ -47,16 +49,19 @@
 class PCollectionVisualizationTest(unittest.TestCase):
 
   def setUp(self):
+    ie.new_env()
     # Allow unit test to run outside of ipython kernel since we don't test the
     # frontend rendering in unit tests.
     pv._pcoll_visualization_ready = True
     # Generally test the logic where notebook is connected to the assumed
     # ipython kernel by forcefully setting notebook check to True.
     ie.current_env()._is_in_notebook = True
 
-    self._p = beam.Pipeline()
+    self._p = beam.Pipeline(ir.InteractiveRunner())
+    ib.watch(self)
     # pylint: disable=range-builtin-not-iterating
     self._pcoll = self._p | 'Create' >> beam.Create(range(1000))
+    ib.watch(self)
```

Comment:
Why call this twice?

Reply:
The first time is to watch the locals including the pipeline `self._p`.
Then during the pipeline construction time, Interactive Beam will implicitly modify the labels such as `Create` if the code is executed in ipython to something like `Cell 1: Create`.

The second watch is to notify Interactive Beam about the PCollections that have been defined so far.
Then when pipeline gets executed, Interactive Beam will cache for PCollections watched in the local scope. The cache will be available even if the `pipeline.run()` or data visualization/introspection is not in the local scope later on.

Reply:
So shouldn't the second call be enough? It should catch locals as well as PCollections, right?

Reply:
The first one has some implicit effects to the transforms applied, so it has to be in-between pipeline creation and transforms.
It might not be obvious for this test, but these 2 lines serves different purposes.